### ISSUE TITLE
PitchTuning is not correct for manual extensions

### ISSUE THREAD

**Issue by larspalo** [GrandOrgue/OdfEdit#41]
Often the SP sample sets have manual extensions (increased compass) that utilize the same samples as a few notes below to provide pipes that might not have existed originally. Currently, with version 2.9, the correct PitchTuning lines are not written for those extensions.

For instance for the St.Michel positif Bourdon 8' has this written from OdfEdit:

```
Pipe049Gain=5.0
Pipe049=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\084-C.wav
Pipe049LoadRelease=N
Pipe049LoopCrossfadeLength=14
Pipe049ReleaseCount=3
Pipe049Release001=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel99999\084-C.wav
Pipe049Release001ReleaseCrossfadeLength=37
Pipe049Release002=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel00190\084-C.wav
Pipe049Release002MaxKeyPressTime=157
Pipe049Release002ReleaseCrossfadeLength=39
Pipe049Release003=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel00390\084-C.wav
Pipe049Release003MaxKeyPressTime=325
Pipe049Release003ReleaseCrossfadeLength=39
Pipe050Gain=2.0
Pipe050=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\080-G#.wav
Pipe050LoadRelease=N
Pipe050LoopCrossfadeLength=15
Pipe050ReleaseCount=3
Pipe050Release001=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel99999\080-G#.wav
Pipe050Release001ReleaseCrossfadeLength=83
Pipe050Release002=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel00190\080-G#.wav
Pipe050Release002MaxKeyPressTime=151
Pipe050Release002ReleaseCrossfadeLength=86
Pipe050Release003=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel00390\080-G#.wav
Pipe050Release003MaxKeyPressTime=322
Pipe050Release003ReleaseCrossfadeLength=86
Pipe051Gain=2.0
Pipe051=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\081-A.wav
Pipe051LoadRelease=N
Pipe051LoopCrossfadeLength=14
Pipe051ReleaseCount=3
Pipe051Release001=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel00390\081-A.wav
Pipe051Release001MaxKeyPressTime=337
Pipe051Release001ReleaseCrossfadeLength=80
Pipe051Release002=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel99999\081-A.wav
Pipe051Release002ReleaseCrossfadeLength=80
Pipe051Release003=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel00190\081-A.wav
Pipe051Release003MaxKeyPressTime=167
Pipe051Release003ReleaseCrossfadeLength=80
Pipe052Gain=2.0
Pipe052=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\082-A#.wav
Pipe052LoadRelease=N
Pipe052LoopCrossfadeLength=15
Pipe052ReleaseCount=3
Pipe052Release001=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel00390\082-A#.wav
Pipe052Release001MaxKeyPressTime=327
Pipe052Release001ReleaseCrossfadeLength=81
Pipe052Release002=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel00190\082-A#.wav
Pipe052Release002MaxKeyPressTime=157
Pipe052Release002ReleaseCrossfadeLength=81
Pipe052Release003=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel99999\082-A#.wav
Pipe052Release003ReleaseCrossfadeLength=83
Pipe053Gain=2.0
Pipe053=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\083-B.wav
Pipe053LoadRelease=N
Pipe053LoopCrossfadeLength=13
Pipe053AttackCount=1
Pipe053Attack001=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\083-B_bis.wav
Pipe053Attack001LoadRelease=N
Pipe053Attack001LoopCrossfadeLength=16
Pipe053ReleaseCount=3
Pipe053Release001=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel00190\083-B.wav
Pipe053Release001MaxKeyPressTime=148
Pipe053Release001ReleaseCrossfadeLength=81
Pipe053Release002=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel00390\083-B.wav
Pipe053Release002MaxKeyPressTime=317
Pipe053Release002ReleaseCrossfadeLength=81
Pipe053Release003=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel99999\083-B.wav
Pipe053Release003ReleaseCrossfadeLength=81
Pipe054Gain=2.0
Pipe054=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\084-C.wav
Pipe054LoadRelease=N
Pipe054LoopCrossfadeLength=19
Pipe054ReleaseCount=3
Pipe054Release001=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel99999\084-C.wav
Pipe054Release001ReleaseCrossfadeLength=83
Pipe054Release002=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel00390\084-C.wav
Pipe054Release002MaxKeyPressTime=332
Pipe054Release002ReleaseCrossfadeLength=81
Pipe054Release003=..\OrganInstallationPackages\000867\pipe_front\pos_Bourdon\rel00190\084-C.wav
Pipe054Release003MaxKeyPressTime=164
Pipe054Release003ReleaseCrossfadeLength=81
```

Where from Pipe050 there is a break back of sample loading that would need a PitchTuning about +500 to match the intende pipe pitch.

I unfortunately know too little about the HW odf format to know what properties need to be linked together to make the proper pitch tuning adjustments, but a few quick searches in the original .xml file reveals that there indeed are possible connections to investigate.

I realize that this is pretty easy to fix afterwards, and the conversion of OdfEdit 2.9 does yield an organ that's playable with a just few warnings about the pitch and the audible break backs at the point of extension (in this case Pipe050 in the manual and Pipe026 in the pedal). But it would be really nice if the conversion did handle this automatically.

**Comment by eturpault** [comment]
I have only the demo sample set of St Michel, it has the stops of the Grand Orgue manual only, and all these stops have 49 pipes, there is no keys extension, so I am unable to reproduce the issue that you are reporting. Could you share me the HW ODF alone [in this folder](https://drive.google.com/drive/folders/1mO9GmHAe8GluRuoq22yzlXm_9-cOVbmS?usp=sharing) so that I can understand and fix the issue ? Thanks.

**Comment by larspalo** [comment]
@eturpault Done. Otherwise, except that small issue, 2.9 seems to work well with this sample set! Thanks!

**Comment by eturpault** [comment]
Thanks, I will look at that.

**Comment by larspalo** [comment]
Perhaps the pitch tuning information is actually missing in the normal way for this sample set. I just checked the information on SP web site https://www.sonusparadisi.cz/en/organs/france/st-michel-en-thierache.html where this section might be relevant:

> Althought the “Original Organ Tuning” is availabe in Hauptwerk, it will not work with the extended compass of the manuals and the pedal. Instead, consider using the dedicated St. Michel temperament chart which will be installed together with the sample set. This temperament chart also features the original tuning of the instrument, while repitching the extended keys perfectly.

If that's indeed the case one would need to not only parse the normal .xml file but the Temperaments folder xml file too. Why I raised this issue was just because of all the warnings that GO will log about the retuning of the extended keys as the Pipe999PitchTuning lines are missing.

**Comment by larspalo** [comment]
@eturpault Testing OdfEdit on St. Carlo Brescia show a somewhat related issue. The extended (or perhaps better said, extrapolated) keys for 037-C#, 039-D#, 042-F# and 044-G# in the standard multi release version, that use seemingly different samples for those notes, but they are actually the same as neighbor tones which need the Pipe999PitchTuning lines to sound in tune.

Actually, searching in any converted odf produced by OdfEdit lately, I cannot find any PitchTuning lines at all at any level in any of the files. Is something missing? Also, as a side note not directly related to this issue, searching in the OdfEdit source code I see that the PitchTuning seems to be limited at +/- 1200 while it should be +/- 1800.

**Comment by eturpault** [comment]
I compared in the HW ODF of St Michel the attributes of the pipes 49 and 54 for example, and I found nothing permitting to know that the sample file of the pipe 54 (which is the sample of the pipe 49) has to be retuned. My idea, as a workaround, was to recover the MIDI note of the sample file (in its file name or in its metadata), and compare it with the MIDI note of the pipe, which is always defined in the HW ODF, and in case of difference to apply a PitchTuning value for this pipe.
But you are right the temperament xml file should be a better solution, I have to see which data are defined inside. Could you please share in the same Google Drive folder the temperament xml file of St Michel ?

**Comment by larspalo** [comment]
> Could you please share in the same Google Drive folder the temperament xml file of St Michel ?

Already done! Thanks for investigating!

**Comment by eturpault** [comment]
In the check of the ODF data I had indeed +/- 1200 as limits for PitchTuning, it was an error, I have changed it to +/- 1800 in the code and in the help file.
Concerning the attribute PitchTuning, I set it only if there is a difference of pitch value in the HW ODF between the pipe section (attribute Pitch_OriginalOrgan_PitchHz) and its linked sample section (attribute Pitch_ExactSamplePitch).
In the St. Carlo Brescia HW ODF the pipe attribute Pitch_OriginalOrgan_PitchHz is never defined, so OdfEdit never sets the PitchTuning in the GO ODF. I will analyze what you are mentioning about some pipes to see what can be improved.

**Comment by eturpault** [comment]
@larspalo : about the extrapolated pipes of St Michel, I don't see how to define their PitchTuning attribute : 
- I cannot rely on the sample file name to get the actual MIDI note of the sample, some sample sets have value in the file name which is not the actual MIDI note of the sample. For example a file named 024-C.wav for a pipe having MIDI note 36.
- I cannot rely neither on the MIDI note defined in the metadata of the sample file, it is not always the actual MIDI note of the sample. For example in St Michel sample set, the file \000869\pipe\GO_burdon8\084-C.wav has MIDI note = 82 in the metadata whereas it is actually a MIDI note 84.
- The HW temperament file contains a table associating frequencies in Hz to MIDI notes, but if I am not wrong in GO ODF it is not possible to set a frequency to a pipe.
Please let me know if something is wrong in my understanding.

About St. Carlo Brescia extrapolated keys for 037-C#, 039-D#, 042-F# and 044-G# in the standard multi release version, for example which pipe number in which rank would need a retuning according to you ?

**Comment by larspalo** [comment]
@eturpault I think that de-coding/understanding how the HW odf format actually handles (any) equivalent to PitchTuning (if any adjustment of pipe/sample pitch in the original temperament does take place and if so, how...) is most interesting for OdfEdit conversions.

The St Michel organ is at a different base pitch, about two semitones lower than current standard, so blind reliance on the MIDI note is not possible for a totally general use case - except when we do have a reference like that provided by the temperament file.

One possible solution to handle the extensions (when the same sample is used for other pipes but offset) without so much HW odf understanding but using the available information in a temperament file would be to compare the actual sample pitch (either written in odf or embedded in the file) to the expected pitch from the temperament file. Of course one must adjust the used MIDI note number in the temperament file to what harmonic number is used for the pipe/sample to get a comparison as it's listed for 8'. The difference between the pipe/sample pitch and the temperament table pitch, adjusted to cents, would be the Pipe999PitchTuning to use. This should eliminate the warnings about the re-tunings in GO.

In St. Carlo Brescia standard version(s) all the pipes with 037-C#, 039-D#, 042-F# and 044-G# needs PitchTuning lines with either 100 or -100 (unfortunately it varies). The Pipe018 of QuintaDecima also needs Pipe018PitchTuning=-100. The problem is that I've not yet understood how to detect this directly from the HW odf. So far, I've just manually corrected it, but of course it would be best if the conversion from OdfEdit did handle it!

As another side note, the St. Carlo Brescia original version as converted by OdfEdit is not loadable due to the short octave. The mismatch of Manual lowest MIDI note and the Rank gives the Stops FirstAccessiblePipeLogicalKeyNumber=-3 which of course GO won't accept. This is of course a corner case - but one that needs manual key mapping (similar to what I did with the Norrfjärden sample set) to be correct.

I'm sorry that I currently cannot present a solution to the PitchTuning issue in the conversion, I must investigate it further...

**Comment by eturpault** [comment]
About St. Carlo Brescia and the pitch tuning required on first sharp notes, I don't see as well how to handle it automatically in OdfEdit. For example samples Principale\036-c__bis1.wav of Pipe001 and Principale\037-c#.wav of Pipe002 are sounding at the same pitch, and in the HW ODF I see nothing indicating that a pitch tuning +100 has to be applied to the sample of the MIDI note 37 (I made text comparison of all sections related to the two pipes).
In the features description of the sample set, it is written : 

> In the original version, the lowest octave is "short", meaning that it is only diatonic C-D-E-F-G-A-B-H. In the extended version, the chromatic semitones were added.

It seems that in the extended version (standard ODF) the semitones have not been actually added to all sharp notes of the lowest octave. Only the sample 046-a# has the corrected pitch.

As you reported the issue is the same for the sample QuintaDecima\053-f.wav of Pipe018 which is sounding the same pitch as the sample QuintaDecima\054-f#.wav of Pipe019, and I see nothing in the HW ODF indicating that a pitch tuning -100 has to be applied to the sample of the MIDI note 53.

I will fix the issue that you reported with the original version. There is a wrong management of the logical keys which are not accessible between MIDI note 36 and the first accessible key (MIDI note 40 in this version).

**Comment by eturpault** [comment]
@larspalo, you wrote :
> Why I raised this issue was just because of all the warnings that GO will log about the retuning of the extended keys as the Pipe999PitchTuning lines are missing.

On which criteria GO is able to detect that pitch tuning is missing for a sample ? On pitch value defined in the metadata of the sample file or by making FFT on the audio data of the file ? OdfEdit should use the same criteria to apply the proper pitch tuning if needed.

**Comment by larspalo** [comment]
> On which criteria GO is able to detect that pitch tuning is missing for a sample ?

If the pitch when applying a selected temperament (other than "Original") for a pipe/sample will be more than 600 cent different than what the original temperament has (or rather: is expected to be) - then there will certainly be a warning.

Perhaps the needed pitch tuning could be arrived at by comparing the embedded pitch (of for instance first attack of a pipe) to what pitch that pipe MIDI key (note) pressed to activate the sample would yield when adjusted for its harmonic number. The difference expressed in cents would be the PitchTuning. But overall I think that a better option for OdfEdit would be that we could de-code how the HW mechanism for this actually works. A custom solution like this could of course be added as an option - to allow correcting sample sets that doesn't provide a (HW) fix for it in the "Original" temperament.

Recently I've used this code in GOODF to detect the HarmonicNumber of a pipe:

```
double embeddedPitch = atkFile.getPitchInHz();
double effectivePitch = embeddedPitch * pow(2, (p.pitchTuning / 1200.0)); // this is the pitch the pipe sample will actually sound at taking existing pitch tuning into account
double expectedEightFootPitch = referencePitch * pow(2, ((double)(pipeMIDInote - 69) / 12.0)); // reference is the pitch of an 8' pipe a1 (A4) in the organ plus the MIDI key activating the pipe
double pitchRatio = effectivePitch / expectedEightFootPitch;
int harmonicNbr = round(8.0f * pitchRatio);
```

which of course is another thing - but in a pinch, a similar concept, but somewhat reversed, could be used to arrive at what pitch tuning would be needed. This concept of course depends on that the embedded pitch accurately describes the pitch of the sample...

**Comment by larspalo** [comment]
@eturpault Almost needless to say, if the sample set provides a temperament file like St Michel the expected pitch can be looked up there instead of trying to calculate it.

The expected pitch tuning could be found using:

PitchTuning = 1200 * log(expectedPitch/embeddedPitch) / log(2)

In the St Michel case of the GO Montre 8' 084-C sample that's used for midi note 89 and that 8' reference pitch taken from the temperament table, the calculation would be:

PitchTuning = 1200 * log(1257.68/940.14) / log(2)

which is approximately equal to 503,78 and what I would expect it to be.

**Comment by larspalo** [comment]
@eturpault In the St Carlo odf xml (which is of an older date and HW version) there is in the pipe_soundengine01 sections for the pipes lines like:

<Pitch_Tempered_BaseTuningDeviation>100</Pitch_Tempered_BaseTuningDeviation>

which likely holds the needed PitchTuning value. I've not yet identified how to find the equivalent information in newer versions of the HW xml file. But I imagine that it shouldn't be hard to add a check in OdfEdit if that tag exists and if so set the PitchTuning from it.

**Comment by eturpault** [comment]
@larspalo
In St Carlo, the attribute Pitch_Tempered_BaseTuningDeviation is set at 100 only for noise samples (keys, stops, blower), it is strange to set a returning value for noises. It is set at 7 for all 30 pipes of the Fiffaro stop. So it should not help for our topic. Currently OdfEdit does not read this attribute.

I will make a try to use the embedded MIDI note and pitch values of the samples to set their PitchTuning value for St Carlo, according to what I see for some sample it may work, I will let you know the result. It will increasing the HW to GO conversion time if OdfEdit reads the metadata of all wav files, but if it is necessary there is no choice.

**Comment by larspalo** [comment]
> It will increasing the HW to GO conversion time if OdfEdit reads the metadata of all wav files, but if it is necessary there is no choice.

I'm very sorry that I've not been able to find the exact equivalent for the conversion. I certainly don't wish to have a default PitchTuning that would exactly equal to that of what a selected "Equal temperament" would yield. The very best would be if we could find out what the original HW sample set creator intended for the "Original temperament" and kept that for the converted version too. Maybe, only for the extreme cases, where the conversion would result in GO warnings/errors, it would be best to intervene and apply other logics that bring the actual values into acceptable range. What do you think about it?

**Comment by eturpault** [comment]
Don't be sorry, you are helping me a lot in understanding some concepts and thus improving OdfEdit HW to GO conversion.

I made the test with St Carlo to apply to each pipe a PitchTuning value caculated from the expected pipe MIDI note and the sample MIDI unity note and MIDI pitch fraction (recovered from the metadata of the wav file). It gives good result for the samples which the MIDI note is defined correctly in the metadata (all keys are separated by semi-tones), but in some samples the MIDI unity note being erroneous (for example 67 instead of 36) a PitchTuning is not possible. This method gives bad results with normal sample sets, somes notes are wrong. I will continue to search a way to convert St Carlo with a good result.

**Comment by eturpault** [comment]
@larspalo 
For St Michel you wrote :
> Where from Pipe050 there is a break back of sample loading that would need a PitchTuning about +500 to match the intende pipe pitch.

For St Carlo Brescia you wrote :
> In St. Carlo Brescia standard version(s) all the pipes with 037-C#, 039-D#, 042-F# and 044-G# needs PitchTuning lines with either 100 or -100 (unfortunately it varies). The Pipe018 of QuintaDecima also needs Pipe018PitchTuning=-100. The problem is that I've not yet understood how to detect this directly from the HW odf. So far, I've just manually corrected it, but of course it would be best if the conversion from OdfEdit did handle it!

After a new deep analysis, I don't see in the two HW ODF an attribute value permitting to automatically correct the pitch of these samples, The solution is either to make the correction manually in the ODF as you did, or in the Audio/Midi menu of GO to force a temperament different from the original one, thus GO makes the pitch correction.

I would be interrested to know, if someone has Hauptwerk and could test the St. Carlo sample set, if the notes 37, 39, 42, 44 are actually played at the right pitch without asking to Hauptwerk to apply another temperament. If it is the right pitch which is played, I guess that Hauptwerk compares the expected MIDI note with the embedded MIDI note, and if they are different it applies itself a pitch correction automatically.

I tried a workaround to set PitchTuning by using the MIDI note and pitch values which are embedded in the sample file, but it has bad effect with other sample sets (somes pipes are totally detuned) and it slows down a lot the conversion duration.
So I prefer to let the current conversion logic unchanged in OdfEdit, it works well with a large majority of sample sets.

If you want to explore the HW ODF, once you have loaded it in OdfEdit, you can navigate hierarchically among the sections in the tab "HW objects" at the right of the OdfEdit window. In [this document](https://github.com/GrandOrgue/OdfEdit/blob/main/doc/OdfEdit_HW_objects_chart.pdf) I made a chart of the various sections and their relationship.

**Comment by eturpault** [comment]
The BIS Romsey Abbey sample set has keyboards extension (MIDI notes 92 to 96), and it has in the HW ODF an attribute giving the MIDI note of the each sample, thanks to that OdfEdit can detect a difference between the pipe MIDI note and the sample MIDI note and thus apply automatically a PitchTuning value (I just implemented it), but in St Michel HW ODF there is not such information unfortunately.

**Comment by larspalo** [comment]
@eturpault Thanks for investigating it. I don't have any problem with making the (very minor) corrections manually in the odf. Nor would I really expect OdfEdit to fix things that are not working in the original HW sample set to begin with.

> it has in the HW ODF an attribute giving the MIDI note of the each sample, thanks to that OdfEdit can detect a difference between the pipe MIDI note and the sample MIDI note and thus apply automatically a PitchTuning value (I just implemented it), but in St Michel HW ODF there is not such information unfortunately.

To me it seems like that in the Pipe_SoundEngine01 section, for each sample, there is both a MIDI note and the Harmonic Number listed. That information alone would allow to find out both expected general pitch and what key should be linked with that pipe. In theory it should be possible to detect the cases where GO will warn about excessive pitch differences, but I understand if you think it's slowing down the conversion too much.

**Comment by eturpault** [comment]
The values in the Pipe_SoundEngine01 sections are used to know the expected MIDI note and harmonic number at pipe level.
Here we are trying to know what is the MIDI note at sample level, to apply a pitch tuning to the sample and align it on the pipe MIDI note if needed.

What I can do as a workaround is to add an option in the menu of OdfEdit to ask to check the MIDI note present in the sample file name (**091**-g.wav for example) to detect that a sample pitch tuning is needed compared to the associated pipe MIDI note. This will solve the issue with the extended keys of St Michel manuals and it will not take too much time in the processing.

For St Carlo notes 037-C#, 039-D#, 042-F# and 044-G# this will not work as the sample files names are aligned on the pipe MIDI note and not on the real sound pitch. The alternative workaround could be to use the MIDI note embedded in the sample files (which is aligned on the real sound pitch, compensated with the harmonic number), but as it slows down a lot the convertion speed to recover the metadata, this method would have to be activated also on demand in the menu of OdfEdit if needed.

I will try these two workarounds and let you know if they work.

**Comment by eturpault** [comment]
Using the MIDI note present in the samples files name permits to set correctly the PitchTuning of the extented keys of the St Michel sample set.

Now I am working on the second workaround for St Carlo. 
@larspalo : Have you a formula to get the expected pipe MIDI note from the MIDI unity note and MIDI pitch fraction of the sample file metadata and the harmonic number ?

**Comment by eturpault** [comment]
I think to have found the formula (taking a reference pitch at 440Hz)  : 
- convert the sample MIDI note into a frequency : F1 = 440 x 2^((sample MIDI note - 69) / 12)
- apply the harmonic number to this frequency : F2 = F1 x 8 / harmonic number
- convert the new frequency into MIDI note : pipe MIDI note = 12 x LOG2 (F2 / 440) + 69

-

**Comment by eturpault** [comment]
Here is the ODF of St Carlo Brescia with a PitchTuning value applied on some pipes, using the MIDI note defined **in the metadata of the sample file of each first attack**, to know if a pipe pitch tuning is needed. This method is slowing down the conversion operation as the sample files have to be opened to read the MIDI note in the metadata, then a MIDI note shift has to be done to take into account the pipe harmonic number if different from 8, but it is the only solution to fix the pitch issue of some pipes of this sample set. It will be an option in OdfEdit v2.10.
[Brescia, St.Carlo, multi release.zip](https://github.com/GrandOrgue/OdfEdit/files/14320378/Brescia.St.Carlo.multi.release.zip)

Here is the ODF of St Michel en Thiérache, with a PitchTuning value applied on some pipes (extended keys), using the MIDI note defined **in the name of the sample file of each first attack**, to know if a pipe pitch tuning is needed. The previous method should work for this sample set as well, but this alternative method using files names is really faster. It will be an option in OdfEdit v2.10.
[St. Michel en Thierache, surround.zip](https://github.com/GrandOrgue/OdfEdit/files/14320385/St.Michel.en.Thierache.surround.zip) (I cannot test it as I don't have the surround version, and the demo version doesn't have extended keys)

**Comment by larspalo** [comment]
> This method cannot be used with sample sets having a base pitch different from the standard one because it would retune all the pipes at the standart pitch.

Yes, of course, nothing to worry about. Thanks for your effort with trying to fix the conversion! The tuning issue you mention is exactly why I allowed/require the user to specify the base pitch when trying to auto detect harmonic number in recent GOODF, a feature most useful for compound stops like Mixtures that break back at certain intervals.

I'll have a look at the files and test them in more detail later. For the St. Carlo Brescia with the short octave, there is one piece of the puzzle that's missing and that's the key mapping that need to change too. As it's done now the manuals begin at MIDI key 040 but the stops/pipes begin at 036 - which means that the wrong pipes will be played back for the intended key. The mapping of the manual keys 040, 042 and 044 must change. From the Norrfjärden sample set this is how I did that first in the manual sections:

[Manual001]
Name=Rückpositief
NumberOfLogicalKeys=61
FirstAccessibleKeyLogicalKeyNumber=1
FirstAccessibleKeyMIDINoteNumber=36
NumberOfAccessibleKeys=49
...
MIDIKey036=000
MIDIKey037=000
MIDIKey038=000
MIDIKey039=000
MIDIKey040=036
MIDIKey042=038
MIDIKey044=040

and then in the panel display

[Panel000Element002]
Type=Manual
Manual=001
...
DisplayFirstNote=40
DisplayKeys=45
DisplayKey001=36
DisplayKey003=38
DisplayKey005=40

**Comment by eturpault** [comment]
> For the St. Carlo Brescia with the short octave, there is one piece of the puzzle that's missing and that's the key mapping that need to change too. As it's done now the manuals begin at MIDI key 040 but the stops/pipes begin at 036 - which means that the wrong pipes will be played back for the intended key. The mapping of the manual keys 040, 042 and 044 must change.

This is fixed for OdfEdit v2.10 and in the ODF that I attached. I use Rank001FirstPipeNumber=5 in the Stop section to use as first pipe the Pipe005 which has MIDI note 40.

**Comment by eturpault** [comment]
I let you test the ODF of St Carlo original (with short octave layout) that I have shared yesterday, the keys are all played with the proper pipe, especially the first key plays actually a sample 040 (which is the 5th pipe in the rank).

**Comment by larspalo** [comment]
> I let you test the ODF of St Carlo original (with short octave layout) that I have shared yesterday, the keys are all played with the proper pipe, especially the first key plays actually a sample 040 (which is the 5th pipe in the rank).

Yes, thanks. I downloaded the .zip file, extracted and played the original version in GO. It opens fine without warnings. So far, so good. But the "short octave" is not working as it should in the .organ file you've provided for the original odf.

In your version the 040 key, which is the first key in the manual, does indeed play the 5th pipe of a stop that is the 040-E sample fifth pipe. Which simply it shouldn't do in the short octave. The version you've provided as downloads above does play chromatically correct to the modern standard keyboards - not as the original short octave should do.

I don't exactly know what is the root of the confusion, but what I expect to hear when I play the St.Carlo original short octave "chromatically" down from the 048-C is the following sequence of samples which is "correct" for the short octave:

048-C, 047-B, 046-A#, 045-A, 040-E, 043-G, 038-D, 041-F, 036-C

but what your version play is

048, 047, 046, 045, 044, 043, 042, 041, 040

which is incorrect for a short octave.

However, I can understand if you think this is too much work to identify and implement for the rather few sample sets containing this feature. No hard feelings if you choose to go that route.

**Comment by eturpault** [comment]
Ok now I understand that what you are calling "short octave" is a special definition of notes arrangements, as described [here](https://en.wikipedia.org/wiki/Short_octave), so far I understood that it was simply a shortened chromatic octave. Thanks for the explanation, I learned something today.
It explains the following description of Sonus Paradisi :
> In the original version, the lowest octave is "short", meaning that it is only diatonic C-D-E-F-G-A-B-H.

Maybe in the future I will try to manage the conversion of this special keys mapping, limited to early keyboard instruments, I think to see in the HW ODF where to find the information (links between DivisionInput and KeyboardKey sections), but for the moment I keep the current implementation.

**Comment by larspalo** [comment]
> Maybe in the future I will try to manage the conversion of this special keys mapping, limited to early keyboard instruments, I think to see in the HW ODF where to find the information (links between DivisionInput and KeyboardKey sections), but for the moment I keep the current implementation.

Yes, I think that's a good decision for now. Thanks for your efforts!

**Comment by eturpault** [comment]
@larspalo Finally I have already implemented in OdfEdit the management of the unusual keys to MIDI notes mapping which can be defined in the HW ODF, using as template for the GO ODF what you have put in a previous comment from your Norrfjärden sample set.
Here is the generated ODF (it is still necessary to ask OdfEdit to read the MIDI note value in the samples metadata to apply a pitch tuning to some pipes). Now the real "short octave" is in place !
[Brescia, St.Carlo, original, multi release.organ.zip](https://github.com/GrandOrgue/OdfEdit/files/14380372/Brescia.St.Carlo.original.multi.release.organ.zip)

**Comment by eturpault** [comment]
Implemented in OdfEdit v2.10.
@larspalo please let me know if you have other comments. They are a great help to improve OdfEdit, thanks.

**Comment by larspalo** [comment]
@eturpault Thanks for the new version!

I can confirm that the short octave works perfectly for the original version of St.Carlo both in the "Original temperament" and in any other chosen. Great work!

Still I don't manage to get the other "special" samples/keys that cause issues (like the C#, D#/Eb, F# and G#) in the standard of St.Carlo or the extended keys of St. Michel to work out of the box in the "Original temperament", as soon as another temperament is chosen it works fine with the tuning. I've tried both of the pitch tuning options in the menu but they seem to give the same results as far as I can tell. Which pitch tuning option should I use in the menu for the "Original temperament" to be tuned correctly?

I also noted that the old style combination system present in St.Carlo (where you can prepare another registration with the smaller levers to the right, and activate it with the right pedal) is not working 100% as I think it should. When the pedal is pushed it should exclusively select between the combination set by the left drawstop (lever) and the smaller right lever, I think. Currently the right combination is added to the left and not replacing the left hand drawstop combination.

**Comment by eturpault** [comment]
@larspalo Maybe I misunderstood the sentence that I extracted from the wikipedia page, let me know what you think about it.
About how this combination mechanism is described in the HW ODF, I have to analyze it again, I have a doubt about what I interpreted.

**Comment by larspalo** [comment]
@eturpault I'm not certain, and as I cannot test it in HW it's impossible to say for sure how it was intended to behave. The information in the SP html file about the console where this feature is described doesn't give any conclusive hints either.

![StCarloInfo](https://github.com/GrandOrgue/OdfEdit/assets/6173941/7f66055e-2cf4-400d-8a9e-e03b758e6ff6)

However, the hint that it behaves as a general led me to believe that it would replace the combination set by the hand register levers, but I honestly don't know for sure. Also, I know too little about the HW switch network to be able to say how it would translate to GO.

In the odf I created a long time ago (and updated today with some small adjustments...) I've implemented the lomb combination with XOR stops and exclusive selection from the stop levers. This was of course not done from the original HW odf as I didn't have any means to do so at that time (it was long before OdfEdit was created). Nor is it certain that this is the correct way, but I attach it here just for reference if it's of any interest to you. I've also placed it differently from the HW odfs - so it should be placed next to your SilentLoop.wav file... [StCarloBrescia.zip](https://github.com/GrandOrgue/OdfEdit/files/14667423/StCarloBrescia.zip)

**Comment by eturpault** [comment]
Looking again in detail at the HW ODF, I confirm that the stops selected by the small levers D are added to the stops selected by the standard levers C when the right pedal B is engaged. So the logical formula to select a rank is : C or (B and D). This is what OdfEdit applies.

This is confirmed by this video : https://www.youtube.com/watch?v=wMRogSDHcW8 in which the Italian organist says at 1:25 (I speak Italian) : 

> "la combinazione alla lombarde consentiva di inserire **in aggiunta** ai registri preparati degli altri disposti in precedenza"

 which means : "the Lombard combination allowed to insert **in addition** to the prepared registers other registers prepared in advance". So there is addition, not substitution when the B pedal is engaged.
The organ of Sandro Carnelos shown in this video has a mechanism different than St Carlo to prepare the stops to add when the B pedal is engaged (by pulling on the stop lever). The one of St Carlo is more convenient to use I think.

Other website saying the same : https://www.campanologia.it/contenuto/pagine/01-ATS/ATS-B02/ATS-B02-02-glos-organo.htm

> Combinazione libera "alla lombarda" (meccanica dell'organo) : è un congegno meccanico a pedale che agisce sulle leve dei tiranti di un certo numero di registri, preventivamente selezionati dall'organista, inserendo le file di canne corrispondenti.

which means : "Free combination "alla lombarda" (organ mechanics) : is a pedal-operated mechanical device that acts on the levers of the tie-rods of a certain number of registers, previously selected by the organist, **inserting** the corresponding rows of pipes."

The advantage of the ODF that you built is that it does exactly what you expect (attribute Percussive is missing in several stops so it cannot be loaded as it is by GO).

**Comment by larspalo** [comment]
You're very likely absolutely right about how the lomb combination should work. I've no personal real life experience with real italian organs (which is also why I bought that sample set). Keep the implementation as close to the HW odf as possible, that's the correct strategy.

> The advantage of the ODF that you built is that it does exactly what you expect (attribute Percussive is missing in several stops so it cannot be loaded as it is by GO, maybe you are using GO 3.14.0 where Percussive will become optional).

Yes, good catch! In the updates of GoOdf I'm working on I've implemented the Percussive as GO will use it in the future (now) - a side effect will be that odfs prepared with it will be incompatible with older versions of GO. As I build GO from latest source myself and also frequently use multiple testing versions to test various things I didn't even think about it... :)

**Comment by eturpault** [comment]
It is fixed as much as possible in OdfEdit v2.10, with added menu options ("correct pipes pitch from samples metadata" and "correct pipes pitch from samples file name") to activate before the conversion when the standard conversion algorithm does not permit to set properly a right Pitch Tuning.
